### ISSUE TITLE
docker 1.10 embedded DNS server forwarding container name requests to external DNS when using curl

### ISSUE THREAD

**Issue by xamoc** [docker/docker#20448]
Description of problem:

I use docker 1.10 and the new service discovery using the embedded DNS server. According the section *Docker embedded DNS server* of the [docker networking documentation](https://docs.docker.com/engine/userguide/networking/dockernetworks/), I understand that only unresolved requests are forwarded to the external DNS server. 

Unfortunately, by looking at my DNS server logs, I find out that some dns requests (when using curl for example) to a container name are forwarded to the external dns server defined on the host machine.

For example, with a simple nginx container named "myweb", the command `curl -XGET http://myweb` from another container on the same network produce the following entry in my DNS server logs:

```
18-Feb-2016 16:48:07.324 queries: info: client MY_HOST_IP#47069 (myweb): query: myweb IN AAAA + (MY_DNS_IP)
```

Strange thing is that a query to "myweb" using dig or nslookup does not appear in the DNS logs.

I tried bridge and overlay networks, using docker-compose 1.6.0 or docker directly and have always the same behavior.

I hope my explanations are clear enough

`docker version`:
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

`docker info`:
Containers: 22
 Running: 3
 Paused: 0
 Stopped: 19
Images: 101
Server Version: 1.10.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 294
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: host bridge null
Kernel Version: 3.19.0-25-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.797 GiB
Name: XXXXXXXXXXX
ID: FF3I:AT6Q:XQTR:TO6I:ZUB5:7MD2:X4E7:OFQP:T2CE:27IE:HPXA:5IZI
WARNING: No swap limit support

`uname -a`:
Linux XXXXXXXXXXX 3.19.0-25-generic #26~14.04.1-Ubuntu SMP Fri Jul 24 21:16:20 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Steps to Reproduce:

1. Setup your host to use a DNS server that you can access the queries logs
2. docker network create mynet
3. docker run -d --net=mynet --name myweb nginx
4. docker run -it --rm --net=mynet debian bash

then, inside the container:

1. apt-get update
2. apt-get install dnsutils curl
3. dig myweb
```
; <<>> DiG 9.9.5-9+deb8u5-Debian <<>> myweb
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 50747
;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;myweb.				IN	A

;; ANSWER SECTION:
myweb.			1800	IN	A	172.19.0.4

;; Query time: 0 msec
;; SERVER: 127.0.0.11#53(127.0.0.11)
;; WHEN: Thu Feb 18 16:03:02 UTC 2016
;; MSG SIZE  rcvd: 44
```

4. curl -XGET http://myweb
```
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

When using dig, the request seems not forwarded but when using curl, the following entry appears in my dns log:

```
18-Feb-2016 16:48:07.324 queries: info: client MY_HOST_IP#47069 (myweb): query: myweb IN AAAA + (MY_DNS_IP)
```

**Comment by sanimej** [comment]
@xamoc The docker embedded DNS server today handles only A records and not AAAA records. If you have IPv6 enabled, some of the tools (nc, curl etc.) first try to resolve the name to IPv6 (AAAA query). Since the embedded server can't handle it, it will be forwarded to the external servers. If you want to avoid this behavior you can use the option to resolve the name to IPv4 only (nc -4, curl -4).

**Comment by xamoc** [comment]
thanks for this quick response @sanimej

The problem is that IPv6 is not enable on my docker daemon (I use the default ubuntu configuration) and I don't know how to force some applications of my stack to resolve on IPv4 only (curl was just an example).

I think I'll have to stick with a dns cache on my host for the moment.

**Comment by xamoc** [comment]
Hi, 

Some additional information: I deactivate IPv6 support on the ubuntu host but nothing change.

@sanimej, do you know when the IPv6 support will be added to the embedded DNS ?

**Comment by sanimej** [comment]
@xamoc I have a PR open to add v6 service discovery in Docker [libnetwork #870] (https://github.com/docker/libnetwork/pull/870). This should be available in the next release, 1.11
